### PR TITLE
[android-client-build] - Add create android client build form.

### PR DESCRIPTION
**Pull request by mikenairn**
Re-opened due to base branch change see https://github.com/aerogear/origin-web-console/pull/41

* Create jenkins pipeline build config for android applications.
* Enable "Create Build" option only when mobile CI/CD provisioned and client type is Android

**Verification:**
- [ ] Create project with Mobile CI/CD provisioned and an android client
- [ ] Select "Create Build" from the android clients action menu
- [ ] Create a new build for a public repo, debug build type, start the pipeline and verify it completes
- [ ] Create a new build for a private repo, debug build type, start the pipeline and verify it completes
- [ ] Create a new build for a public repo, release build type, start the pipeline and verify it completes
- [ ] Create a new build for a private repo, release build type, start the pipeline and verify it completes

To create a test keystore certificate, use the following:

```
keytool -genkey -v -keystore testcert.keystore -alias testcert -keyalg RSA -keysize 2048 -validity 10000
keytool -importkeystore -srckeystore testcert.keystore -destkeystore testcert.p12 -deststoretype PKCS12 -srcalias testcert
```

**Notes:**
* Test application with a working Jenkinsfile can be found in this repo https://github.com/mikenairn/helloworld-android-gradle/tree/5xjenkinsfile

### CODE REVIEW

**Review by sedroche — COMMENTED** [review]
LGTM @mikenairn

**Review by vchepeli — APPROVED** [review]

**Comment by vchepeli** [comment]
LGTM @mikenairn

**Comment by JameelB** [comment]
**Verified**

- [x] Verified that the `Create Build` option under the actions menu of a mobile client is only available when it is an android client and when mobile ci|cd is provisioned.
- [x] Successfully completed a build for a public repo with a build type of debug.
- [x] Successfully completed a build for a public repo with a build type of release.
- [x] Successfully completed a build for a private repo with a build type of debug.
- [x] Successfully completed a build for a private repo with a build type of release.

@sedroche @mikenairn I noticed that the `Create Build` option is available even when the mobile ci|cd provision hasn't finished yet. I think this should be disabled until the service has finished provision because builds created before that will fail. wdyt?

I've also noticed that when an error occurs during build creation, it seems that a build is still created. These errors mostly occur due to duplicate secret names for the git ssh key/keystore cert. The form doesn't exit out to the client view and stays on the same page. Is this a known behaviour?
image
image

**Comment by sedroche** [comment]
@mikenairn There is a filter called `isServiceInstanceReady` to check for when the service is finished provisioning.
We will need to add some validation of secret and build config names it seems

**Comment by mikenairn** [comment]
Thanks @JameelB 

> I noticed that the Create Build option is available even when the mobile ci|cd provision hasn't finished yet.

Yes you are right, had meant to come back to this. Will update the test to use the filter @sedroche described above.

> I've also noticed that when an error occurs during build creation, it seems that a build is still created.

The way this form is working is a bit wrong tbh, in that it is creating multiple resources with one create action. Instead of doing this, we should be using an approach seen on other forms where we instead ask the user to select an existing secret, or they are moved off this form to another modal view to create a new secret. This way errors would be handled as expected and the user can deal with the errors at the correct time.

I think making these changes is probably better than trying to hack in validation as it currently is, so will create a follow on issue for that. This was deemed low priority before, but considering it can cause issues with resources being created when they shouldn't, we could maybe change that.

**Comment by JameelB** [comment]
@mikenairn Creating a follow up issue for this sounds good to me. I'll review it again once the filter is used :+1:

File: app/scripts/controllers/mobileClients.js [hunk 1]
```diff
@@ -73,6 +76,15 @@ angular.module('openshiftConsole')
                 }
                 ctrl.mobileClient = mobileClient;
               }));
+
+              watches.push(DataService.watch(serviceInstanceVersion, context, function (serviceInstances) {
+                $scope.serviceInstances = serviceInstances.by("metadata.name");
+                ctrl.mobileCIEnabled = _.some($scope.serviceInstances, function(serviceInstance) {
+                  console.log(serviceInstance.metadata.name + " :: ready = " + isServiceInstanceReady(serviceInstance));
+                  return /aerogear-digger/.test(serviceInstance.metadata.name) && isServiceInstanceReady(serviceInstance);
```
Comment (mikenairn) [inline app/scripts/controllers/mobileClients.js:84]: @sedroche @JameelB Filtered now to avoid setting mobileCIEnabled before the service is ready.

**Comment by mikenairn** [comment]
Couple of follow on issues created:
* https://issues.jboss.org/browse/AEROGEAR-2815
* https://issues.jboss.org/browse/AEROGEAR-2814

**Review by maleck13 — COMMENTED** [review]
With 3 ppl already reviewing I will leave it to them

**Pull request merged**